src/V3Timing.cpp

```diff
@@ -85,18 +90,144 @@ class TimingVisitor final : public VNVisitor {
     };
 
     // NODE STATE
-    //  AstNode::user1()                         -> bool.         Set true if the node has been
+    //  AstClass::user1()                        -> bool.         Set true if the class member
+    //                                                            cache has been refreshed.
+    //  Ast{NodeProcedure,CFunc,Begin}::user2()  -> bool.         Set true if process/task is
+    //                                                            suspendable
+    //  Ast{NodeProcedure,CFunc,Begin}::user3()  -> DependencyVertex*.  Vertex in m_depGraph
+    const VNUser1InUse m_user1InUse;
+    const VNUser2InUse m_user2InUse;
+    const VNUser3InUse m_user3InUse;
+
+    // STATE
+    AstClass* m_classp = nullptr;  // Current class
+    AstNode* m_procp = nullptr;  // NodeProcedure/CFunc/Begin we're under
+    V3Graph m_depGraph;  // Dependency graph where a node is a dependency of another if it being
+                         // suspendable makes the other node suspendable
+
+    // METHODS
+    // Get or create the dependency vertex for the given node
+    DependencyVertex* getDependencyVertex(AstNode* const nodep) {
+        if (!nodep->user3p()) nodep->user3p(new DependencyVertex{&m_depGraph, nodep});
+        return nodep->user3u().to<DependencyVertex*>();
+    }
+    // Create a dependency edge between the given nodes
+    void makeDependencyEdge(DependencyVertex* const fromVxp, DependencyVertex* const toVxp) {
+        for (V3GraphEdge* edgep = fromVxp->outBeginp(); edgep; edgep = edgep->inNextp()) {
+            if (edgep->top() == toVxp) return;
+        }
+        new V3GraphEdge{&m_depGraph, fromVxp, toVxp, 1};
+    }
+    // Propagate suspendable flag to all nodes that depend on the given one
+    void propagateSuspendable(DependencyVertex* const vxp) {
```

Comment:
Are you certain all the child vertices will be in place at the time this is called as each vertex was built?  Seems safer to instead have the entire graph be built, then this done on the graph as a whole.

Reply:
Newly visited nodes can directly refer to the stuff they depend on to see if they're suspendable, i.e. if we're in a process, we can just see if it calls any suspendable functions.

The graph is used for revisiting previously handled nodes (let's say a function turns into a coroutine; we have to go back to processes that call it and make those coroutines too), and for non-direct relationships (a method overrides another method).

This mostly made sense when everything was happening in one visitor. I guess after the split we may as well just first build the graph, then propagate. I'll try that.

Reply:
It's more intensive to remove edges then just follow the extra edges, so if the only reason for deduplicating was to avoid traversal, just add the duplicates.

Reply:
I removed this deduplication, but added a `removeTransitiveEdges` call before propagation. Doesn't seem to significantly affect performance on VeeR and UVM, but the debug graph is cleaner.